### PR TITLE
Docs page on internal design

### PR DESCRIPTION
**Pull request by TomNicholas**
Adds a new page to the xarray internals documentation giving an overview of the internal design of xarray.

This should be helpful for xarray contributors and for developers of extensions because nowhere in the docs does it really explain how `DataArray` and `Dataset` are constructed from `Variable`.

- [ ] ~~Closes #xxxx~~
- [ ] ~~Tests added~~
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] ~~New functions/methods are listed in `api.rst`~~

### CODE REVIEW

File: doc/internals/variable-objects.rst [hunk 1]
```diff
@@ -1,31 +0,0 @@
-Variable objects
```
Comment (TomNicholas) [inline doc/internals/variable-objects.rst:1]: All of this content has been kept in `internal-design.rst`. I'm not sure why `git mv` didn't seem to understand that I was just renaming this file.

File: doc/internals/internal-design.rst [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    var._data
+
+You're looking at one of xarray's internal `Lazy Indexing Classes`. These powerful classes are hidden from the user,
+but provide important functionality.
+
+Calling the public :py:attr:`~xarray.Variable.data` property loads the underlying array into memory.
+
+.. ipython:: python
+
+    var.data
```
Comment (TomNicholas) [inline doc/internals/internal-design.rst:177]: These examples print long reprs of the numpy arrays, but irritatingly `np.set_printoptions(threshold=10)` doesn't seem to do anything when I try it locally.

File: doc/internals/internal-design.rst [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+A :py:class:`~xarray.DataArray` is a composite object consisting of multiple
+:py:class:`~xarray.core.variable.Variable` objects which store related data.
+
+A single :py:class:`~xarray.core.Variable` is referred to as the "data variable", and stored under the :py:attr:`~xarray.DataArray.variable`` attribute.
+A :py:class:`~xarray.DataArray` inherits all of the properties of this data variable, i.e. ``dims``, ``data``, ``attrs`` and ``encoding``,
+all of which are implemented by forwarding on to the underlying ``Variable`` object.
+
+In addition, a :py:class:`~xarray.DataArray` stores additional ``Variable`` objects stored in a dict under the private ``_coords`` attribute,
+each of which is referred to as a "Coordinate Variable". These coordinate variable objects are only allowed to have ``dims`` that are a subset of the data variable's ``dims``,
+and each dim has a specific length. This means that the full :py:attr:`~xarray.DataArray.size` of the dataarray can be represented by a dictionary mapping dimension names to integer sizes.
```
Comment (headtr1ck) [inline doc/internals/internal-design.rst:98]:
```suggestion
and each dim has a specific length. This means that the full :py:attr:`~xarray.DataArray.sizes` of the dataarray can be represented by a dictionary mapping dimension names to integer sizes.
```

**Review by headtr1ck — COMMENTED** [review]
That's a good addition!

**Comment by TomNicholas** [comment]
@pydata/xarray any other opinions on this? Thoughts on extra parts of xarray's internals that might be useful to explain?

File: doc/internals/internal-design.rst [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+.. _internal design.dataset:
+
+Dataset Objects
+~~~~~~~~~~~~~~~
+
+The :py:class:`~xarray.Dataset` class is a generalization of the :py:class:`~xarray.DataArray` class that can hold multiple data variables.
+Internally all data variables and coordinate variables are stored under a single ``variables`` dict, and coordinates are
+specified by storing their names in a private ``_coord_names`` dict.
+
+The dataset's ``dims`` are the set of all dims present across any variable, but (similar to in dataarrays) coordinate
```
Comment (headtr1ck) [inline doc/internals/internal-design.rst:127]: Unfortunately the `dims` attribute is a dict. Maybe better simply calling it `dimensions` here?

File: doc/internals/index.rst [hunk 5]
```diff
@@ -11,15 +11,14 @@ compiled code to :ref:`optional dependencies<installing>`.
 The pages in this section are intended for:
 
 * Contributors to xarray who wish to better understand some of the internals,
-* Developers who wish to extend xarray with domain-specific logic, perhaps to support a new scientific community of users,
-* Developers who wish to interface xarray with their existing tooling, e.g. by creating a plugin for reading a new file format, or wrapping a custom array type.
-
+* Developers from other fields who wish to extend xarray with domain-specific logic, perhaps to support a new scientific community of users,
+* Developers of other packages who wish to interface xarray with their existing tools, e.g. by creating a plugin for reading a new file format, or wrapping a custom array type.
```
Comment (headtr1ck) [inline doc/internals/index.rst:15]:
```suggestion
* Developers of other packages who wish to interface xarray with their existing tools, e.g. by creating a backend for reading a new file format, or wrapping a custom array type.
```

**Review by headtr1ck — APPROVED** [review]
Since #7368 Coordinates is now public API, but still used internally. Not sure if there is much to explain though.

File: doc/internals/internal-design.rst (hunk 4 above)
Comment (TomNicholas) [inline doc/internals/internal-design.rst:127]:
```suggestion
The dataset's dimensions are the set of all dims present across any variable, but (similar to in dataarrays) coordinate
```

**Pull request merged**